// capnp-ts/lib-test/integration/foo-new.capnp.js
/**
 * This file has been automatically generated by the [capnpc-ts utility](https://github.com/jdiaz5513/capnp-ts).
 */
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "tslib", "../../lib/index"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var tslib_1 = require("tslib");
    var index_1 = require("../../lib/index");
    exports._capnpFileId = "e0b7ff464fbc7ee1";
    var Foo = /** @class */ (function (_super) {
        tslib_1.__extends(Foo, _super);
        function Foo() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        Foo.prototype.getBar = function () { return index_1.Struct.getText(0, this); };
        Foo.prototype.setBar = function (value) { index_1.Struct.setText(0, value, this); };
        Foo.prototype.getBaz = function () { return index_1.Struct.getText(1, this); };
        Foo.prototype.setBaz = function (value) { index_1.Struct.setText(1, value, this); };
        Foo.prototype.toString = function () { return "Foo_" + _super.prototype.toString.call(this); };
        Foo._capnp = { displayName: "Foo", id: "9e8e6186e9348688", size: new index_1.ObjectSize(0, 2) };
        return Foo;
    }(index_1.Struct));
    exports.Foo = Foo;
});
