// @commaai/pandajs/lib/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SAFETY_HONDA_BOSCH_HARNESS = exports.SAFETY_NOOUTPUT = exports.SAFETY_GM_ASCM = exports.SAFETY_ALLOUTPUT = exports.SAFETY_TOYOTA_IPAS = exports.SAFETY_VOLKSWAGEN = exports.SAFETY_MAZDA = exports.SAFETY_SUBARU = exports.SAFETY_TESLA = exports.SAFETY_CHRYSLER = exports.SAFETY_HYUNDAI = exports.SAFETY_CADILLAC = exports.SAFETY_FORD = exports.SAFETY_HONDA_BOSCH_GIRAFFE = exports.SAFETY_GM = exports.SAFETY_ELM327 = exports.SAFETY_TOYOTA = exports.SAFETY_HONDA_NIDEC = exports.SAFETY_SILENT = undefined;
exports.default = Panda;

var _pandaDevice = require('./panda-device');

var _pandaDevice2 = _interopRequireDefault(_pandaDevice);

var _panda = require('./panda');

var _panda2 = _interopRequireDefault(_panda);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var SAFETY_SILENT = exports.SAFETY_SILENT = 0;
var SAFETY_HONDA_NIDEC = exports.SAFETY_HONDA_NIDEC = 1;
var SAFETY_TOYOTA = exports.SAFETY_TOYOTA = 2;
var SAFETY_ELM327 = exports.SAFETY_ELM327 = 3;
var SAFETY_GM = exports.SAFETY_GM = 4;
var SAFETY_HONDA_BOSCH_GIRAFFE = exports.SAFETY_HONDA_BOSCH_GIRAFFE = 5;
var SAFETY_FORD = exports.SAFETY_FORD = 6;
var SAFETY_CADILLAC = exports.SAFETY_CADILLAC = 7;
var SAFETY_HYUNDAI = exports.SAFETY_HYUNDAI = 8;
var SAFETY_CHRYSLER = exports.SAFETY_CHRYSLER = 9;
var SAFETY_TESLA = exports.SAFETY_TESLA = 10;
var SAFETY_SUBARU = exports.SAFETY_SUBARU = 11;
var SAFETY_MAZDA = exports.SAFETY_MAZDA = 13;
var SAFETY_VOLKSWAGEN = exports.SAFETY_VOLKSWAGEN = 15;
var SAFETY_TOYOTA_IPAS = exports.SAFETY_TOYOTA_IPAS = 16;
var SAFETY_ALLOUTPUT = exports.SAFETY_ALLOUTPUT = 17;
var SAFETY_GM_ASCM = exports.SAFETY_GM_ASCM = 18;
var SAFETY_NOOUTPUT = exports.SAFETY_NOOUTPUT = 19;
var SAFETY_HONDA_BOSCH_HARNESS = exports.SAFETY_HONDA_BOSCH_HARNESS = 20;

function Panda(options) {
  options = options || {};
  options.selectDevice = options.selectDevice || selectFirstDevice;

  var device = new _pandaDevice2.default(options);
  options.device = device;
  return new _panda2.default(options);
}

function selectFirstDevice(devices) {
  return devices[0];
}
